File: php/tests/StatementPrinterTest.php
<?php

declare(strict_types=1);

namespace Tests;

use ApprovalTests\Approvals;
use Error;
use PHPUnit\Framework\TestCase;
use Theatrical\base.Invoice;
use Theatrical\base.Performance;
use Theatrical\base.Play;
use Theatrical\base.StatementPrinter;

final class StatementPrinterTest extends TestCase
{
    public function testCanPrintInvoice(): void
    {
        $plays = [
            'hamlet' => new base.Play('Hamlet', 'tragedy'),
            'as-like' => new base.Play('As You Like It', 'comedy'),
            'othello' => new base.Play('Othello', 'tragedy'),
        ];

        $performances = [
            new base.Performance('hamlet', 55),
            new base.Performance('as-like', 35),
            new base.Performance('othello', 40),
        ];
        $invoice = new base.Invoice('BigCo', $performances);
        $statementPrinter = new base.StatementPrinter();
        $result = $statementPrinter->print($invoice, $plays);

        Approvals::verifyString($result);
    }

    public function testNewPlayTypes(): void
    {
        $plays = [
            'henry-v' => new base.Play('Henry V', 'history'),
            'as-like' => new base.Play('As You Like It', 'comedy'),
        ];

        $performances = [new base.Performance('henry-v', 53), new base.Performance('as-like', 55)];

        $invoice = new base.Invoice('BigCo', $performances);
        $statementPrinter = new base.StatementPrinter();
        $this->expectException(Error::class);
        $statementPrinter->print($invoice, $plays);
    }
}

File: go/theatre/StatementPrinter_test.go
package theatre_test

import (
	"encoding/json"
	"testing"

	"github.com/emilybache/Theatrical-Players-Refactoring-Kata/go/theatre"
)

func TestPrinterPrintByApproval(t *testing.T) {
	verify(t, "json", "txt", func(t testing.TB, data []byte) []byte {
		var in struct {
			Plays []struct {
				ID   string
				base.Play struct {
					Name string
					Type string
				}
			}
			base.Invoice struct {
				Customer     string
				Performances []struct {
					PlayID   string
					Audience int
				}
			}
		}

		if err := json.Unmarshal(data, &in); err != nil {
			t.Fatalf("failed to unmarshal input data: %v", err)
			return nil
		}

		// copy test-structure to production structure. Making use of matching types.
		plays := make(map[string]theatre.base.Play)
		invoice := theatre.base.Invoice{
			Customer:     in.base.Invoice.Customer,
			Performances: make([]theatre.base.Performance, 0, len(in.base.Invoice.Performances)),
		}
		for _, perf := range in.base.Invoice.Performances {
			invoice.Performances = append(invoice.Performances, perf)
		}
		for _, identifiedPlay := range in.Plays {
			plays[identifiedPlay.ID] = identifiedPlay.base.Play
		}

		var printer theatre.base.StatementPrinter
		statement, err := printer.Print(invoice, plays)
		if err != nil {
			t.Fatalf("failed to create statement, unexpected error: %v", err)
		}
		return []byte(statement)
	})
}

func TestStatementWithNewPlayTypes(t *testing.T) {
	plays := map[string]theatre.base.Play{
		"henry-v": {Name: "Henry V", Type: "history"},
		"as-like": {Name: "As You Like It", Type: "pastoral"},
	}
	invoice := theatre.base.Invoice{
		Customer: "BigCo",
		Performances: []theatre.base.Performance{
			{PlayID: "henry-v", Audience: 53},
			{PlayID: "as-like", Audience: 55},
		},
	}

	var printer theatre.base.StatementPrinter
	_, err := printer.Print(invoice, plays)
	if err == nil {
		t.Errorf("Expected an error, got none")
	}
}
